src/libraries/FeeLogic.sol

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.0;
+
+import {SafeCast} from 'openzeppelin-contracts/contracts/utils/math/SafeCast.sol';
+import {SafeERC20, IERC20, Address} from 'openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol';
+import {IAllowanceTransfer} from 'permit2/interfaces/IAllowanceTransfer.sol';
+import {IParam} from 'src/interfaces/IParam.sol';
+
+library FeeLogic {
+    using Address for address payable;
+    using SafeCast for uint256;
+    using SafeERC20 for IERC20;
+
+    address internal constant NATIVE = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+    uint256 internal constant BPS_BASE = 10_000;
+
+    event FeeCharged(address indexed token, uint256 amount, bytes32 metadata);
+
+    error InvalidChargeTarget();
+
+    function charge(IParam.Fee memory fee, address feeCollector) internal {
+        address token = fee.token;
+        uint256 amount = fee.amount;
+        if (amount == 0) {
+            return;
+        } else if (token == NATIVE) {
+            payable(feeCollector).sendValue(amount);
+        } else {
+            IERC20(token).safeTransfer(feeCollector, amount);
+        }
+
+        emit FeeCharged(token, amount, fee.metadata);
+    }
+
+    function chargeFrom(IParam.Fee memory fee, address feeCollector, address from, address permit2) internal {
+        address token = fee.token;
+        uint256 amount = fee.amount;
+        if (amount == 0) {
+            return;
+        } else if (token == NATIVE) {
+            revert InvalidChargeTarget();
```

Comment:
I personally think that since it is extracted to a library, it itself needs to be robust enough to be used, which means that it needs to do necessary error handling. If we must remove it for the consideration of gas consumption or other reason, at least we should leave a note for `@dev`.

Reply:
I think `NATIVE` behaves same as any other non-ERC20 tokens in `chargeFrom` of this library, leading to Permit2 reverts, unless we want to distinguish native error from non-ERC20 tokens.

Reply:
Native token is obviously more possible to be applied as fee comparing to other non-ERC20 tokens. I'll just remove it and leave a note.